go/ai/prompt.go

```diff
@@ -426,3 +430,149 @@ func renderDotprompt(templateText string, variables map[string]any, defaultInput
+}
+
+// LoadPromptFolderRecursively recursively loads prompts and partials from the directory.
+func LoadPromptFolderRecursively(registry *registry.Registry, dir string, ns string, subDir string) {
+	parentPath := filepath.Join(dir, subDir)
+	entries, err := os.ReadDir(parentPath)
+	if err != nil {
+		log.Printf("Failed to read directory: %v", err)
+		return
+	}
+
+	for _, entry := range entries {
+		fileName := entry.Name()
+		if entry.IsDir() {
+			LoadPromptFolderRecursively(registry, dir, ns, filepath.Join(subDir, fileName))
+		} else if strings.HasSuffix(fileName, ".prompt") {
+			if strings.HasPrefix(fileName, "_") {
+				partialName := strings.TrimSuffix(fileName[1:], ".prompt")
+				source, err := os.ReadFile(filepath.Join(parentPath, fileName))
+				if err != nil {
+					log.Printf("Failed to read partial file: %v", err)
+					continue
+				}
+				definePartial(registry, partialName, string(source))
+				log.Printf("Registered Dotprompt partial \"%s\" from \"%s\"", partialName, filepath.Join(parentPath, fileName))
+			} else {
+				loadPrompt(registry, dir, fileName, subDir, ns)
+			}
+		}
+	}
+}
+
+// definePartial registers a partial template in the registry.
+func definePartial(registry *registry.Registry, name string, source string) {
+	// TODO: Add this functionality
+}
+
+// loadPrompt loads a single prompt into the registry.
+func loadPrompt(registry *registry.Registry, path string, filename string, prefix string, ns string) {
+	name := strings.TrimSuffix(filename, ".prompt")
+	var variant string
+	if strings.Contains(name, ".") {
+		parts := strings.SplitN(name, ".", 2)
+		name = parts[0]
+		variant = parts[1]
+	}
+
+	source, err := os.ReadFile(filepath.Join(path, prefix, filename))
+	if err != nil {
+		log.Printf("Failed to read prompt file: %v", err)
+		return
+	}
+	parsedPrompt, err := registry.Dotprompt.Parse(string(source))
+	if err != nil {
+		log.Printf("Failed to parse prompt: %v", err)
+		return
+	}
+	metadata, err := registry.Dotprompt.RenderMetadata(string(source), &parsedPrompt.PromptMetadata)
+	if err != nil {
+		log.Printf("Failed to render metadata: %v", err)
+		return
+	}
+
+	commonOpts := commonOptions{
+		ModelName: metadata.Model,
+	}
+	if toolChoice, ok := metadata.Raw["toolChoice"].(ToolChoice); ok {
+		commonOpts.ToolChoice = toolChoice
+	}
+
+	if maxTurns, ok := metadata.Raw["maxTurns"].(int); !ok {
+		commonOpts.MaxTurns = maxTurns
+	}
+
+	if returnToolRequests, ok := metadata.Raw["returnToolRequests"].(bool); !ok {
+		commonOpts.ReturnToolRequests = returnToolRequests
+		commonOpts.IsReturnToolRequestsSet = true
+	}
+
+	// Add the variant to the metadata if it exists
+	if variant != "" {
+		metadata.Variant = variant
+	}
+
+	promptOpt := promptOptions{
+		commonOptions: commonOpts,
+		DefaultInput:  metadata.Input.Default,
+	}
+
+	if inputSchema, ok := metadata.Input.Schema.(*jsonschema.Schema); ok {
+		promptOpt.InputSchema = inputSchema
+	}
+
+	outputSchemaMap := map[string]any{}
```

Comment:
Just noticed, this isn't quite right when combined with `WithOutputType`. `WithOutputType` will reflect the type into a JSON schema. You will want to create the options struct manually here as well like with input.